### ISSUE TITLE
Research usage of Amazon S3 as cloud storage

### ISSUE THREAD

**Issue by scottnath** [punchcard-cms/punchcard#460]
Need to experiment with Amazon S3

```gherkin
Feature: Research S3
  As a core developer
  I want an understanding of Amazon's S3 Cloud Storage
  So that I can determine a level of effort for integrating S3 into Punchcard

  Scenario: Full On Understanding
    Given that I need to have external file storage for images
      And that Amazon S3 is our top contender for cloud storage
     When I experiment with how S3 works
     Then I need to see the level of effort it will require to save files to S3
      And I need to understand what kind of settings would be required in punchcard
      And I need to understand how to set up any keys/secrets/ssh that may be required for connecting to S3
      And I need understand how to integrate the free option into punchcard
      And I need to document my research
      And I need to document my plan and ideas for integrating with S3
```

---

**Research S3**
- [x] Full On Understanding

**Comment by scottnath** [comment]
My understanding is now `full on`.

## Amazon S3

it's a data store. images, objects, files, whatever we like really. We'll be using it for files (which includes images).

### Setup

#### Users & keys & secrets - Oh my!
The setup for this is on the individual or group. We cannot create a public Amazon S3 for punchcard and allow that to be used by the public in general. We _can_ have the required setup documented and we can use this setup to add a `deploy to heroku` button, which would then request AWS key/secret during deployment.

For use on an actual project, we'll need to have a project AWS account, which would require a credit card. This account could create a group, then individual users could have their own key/secret (to use locally.) The project itself would have it's own, non-specific-person user with it's own key/secret.

#### Buckets...or where does my things goes?

You put your stuff in [buckets](http://docs.aws.amazon.com/AmazonS3/latest/dev/UsingBucket.html) - which are what they sound like, separate buckets of storage space.

Files are managed via REST requests. ["Making Requests" dev guide](http://docs.aws.amazon.com/AmazonS3/latest/dev/MakingRequests.html)

### Heroku

Because we're using Heroku for hosting, I needed to learn how S3 connects to it.

To get this going, I needed to figure out environment variables - which made me need to learn how Heroku does this. 

Check out this [great, excellent tutorial](https://devcenter.heroku.com/articles/getting-started-with-nodejs) on Heroku and node running through a field of daisies together.

In short, for local development, we'll need to have a `.env` file setup. For viewing on Heroku, you'll use the [Heroku CLI](https://devcenter.heroku.com/articles/heroku-command-line) to add Heroku [config variables](https://devcenter.heroku.com/articles/config-vars) with `config:set`.
